test/javaTests/src/org/criu/java/tests/CheckpointRestore.java

```diff
@@ -0,0 +1,345 @@
+		try {
+		/*
+		 * We are exiting the test process with exit code 5 in case of an exception
+		 */
+		exitCode = process.exitValue();
+		/*
+		 * Reaching here implies that .exitValue() has not thrown an exception, so the process has
+		 * exited, We now check the exitCode.
+		 */
+		if (5 == exitCode) {
+			Assert.fail(testName + ": Exception occurred while running the test: check the log file for details.");
+		} else {
+			Assert.fail(testName + ": ERROR: Unexpected value of exit code: " + exitCode + ", expected: 5");
+		}
+		} catch (IllegalThreadStateException e) {
+		/*
+		 * Do nothing, as an Exception is expected if the process has not exited
+		 * and we try to get its exitValue.
+		 */
+		}
+
+		currentState = mappedByteBuffer.getChar(Helper.MAPPED_INDEX);
+	}
+
+	/*
+	 * Mapped Buffer state should be Helper.STATE_CHECKPOINT for checkpointing or Helper.STATE_END if some error occurs in test
+	 */
+	if (Helper.STATE_END != currentState) {
+		Assert.assertEquals(currentState, Helper.STATE_CHECKPOINT, testName + ": ERROR: Error occurred while running the test: test is not in the excepted 'waiting to be checkpointed state': " + currentState);
+	} else {
+		Assert.fail(testName + ": ERROR: Error took place in the test check the log file for more details");
+	}
+	/*
+	 * Reaching here implies that MappedByteBuffer is in To Be Checkpointed state.
+	 * Get the pid of the test process
+	 */
+
+	pid = getPid(testName);
+	try {
+		/*
+		 * Checkpoint and Restore the process, the checkpoint function
+		 * calls the Restore function.
+		 */
+		checkpoint(pid);
+
+	} catch (Exception e) {
+		/*
+		 * If exception occurs put the MappedByteBuffer to Helper.STATE_TERMINATE-Terminate state.
+		 * On reading the terminate state, the test process terminates, else it
+		 * may go on looping.
+		 */
+		mappedByteBuffer.putChar(Helper.MAPPED_INDEX, Helper.STATE_TERMINATE);
+		Assert.fail(testName + ": Exception occurred while during checkpointing" + e, e);
+	}
+	/*
+	 * Wait for test process to finish
+	 */
+	currentState = mappedByteBuffer.getChar(Helper.MAPPED_INDEX);
+	while (Helper.STATE_RESTORE == currentState) {
+		currentState = mappedByteBuffer.getChar(Helper.MAPPED_INDEX);
+	}
+
+	/*
+	 * If a test passes it puts the MappedByteBuffer to Helper.STATE_PASS-Pass state,
+	 * On failing to Helper.STATE_FAIL-Fail state, and if our Buffer is in Helper.STATE_TERMINATE state
+	 * its because the checkpoint-restore of test process failed.
+	 */
+
+	Assert.assertNotEquals(currentState, Helper.STATE_TERMINATE, testName + ": ERROR: Checkpoint-Restore failed");
+	Assert.assertNotEquals(currentState, Helper.STATE_FAIL, testName + ": ERROR: Test Failed, Check Log for details");
+	Assert.assertEquals(currentState, Helper.STATE_PASS, testName + " ERROR: Unexpected State of Mapped Buffer");
+	System.out.println("-----" + "PASS" + "-----");
+
+	}
+
+	/**
+	 * Remove .img files, stats-dump and stats-restore files from Log Directory
```

Comment:
Actually keeping these files is useful for debugging, for example, the stats-* files contain information about how much time it took for checkpoint/restore and the .img files can be used to inspect the state of a process.

Reply:
Now the .img files, log files stats-dump and stats-restore are saved in a different folder on failure. This folder is deleted the next time these tests are run. However if we do not want the folder to be deleted on the following runs, the -DneverCleanFailures option can be used, as mentioned in README